spring-graphql/src/main/java/org/springframework/graphql/data/QuerydslDataFetcherSupport.java

```diff
@@ -0,0 +1,260 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.graphql.data;
+
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Function;
+
+import com.querydsl.core.types.EntityPath;
+import com.querydsl.core.types.Predicate;
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+
+import org.springframework.core.ResolvableType;
+import org.springframework.core.annotation.MergedAnnotations;
+import org.springframework.core.convert.support.DefaultConversionService;
+import org.springframework.data.mapping.model.EntityInstantiators;
+import org.springframework.data.projection.ProjectionFactory;
+import org.springframework.data.projection.SpelAwareProxyProjectionFactory;
+import org.springframework.data.querydsl.QuerydslPredicateExecutor;
+import org.springframework.data.querydsl.SimpleEntityPathResolver;
+import org.springframework.data.querydsl.binding.QuerydslBinderCustomizer;
+import org.springframework.data.querydsl.binding.QuerydslBindings;
+import org.springframework.data.querydsl.binding.QuerydslPredicateBuilder;
+import org.springframework.data.repository.NoRepositoryBean;
+import org.springframework.data.repository.Repository;
+import org.springframework.data.repository.core.support.DefaultRepositoryMetadata;
+import org.springframework.data.util.ClassTypeInformation;
+import org.springframework.data.util.Streamable;
+import org.springframework.data.util.TypeInformation;
+import org.springframework.util.Assert;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+
+/**
+ * Utility to implement {@link DataFetcher} based on Querydsl {@link Predicate} through {@link QuerydslPredicateExecutor}.
+ * Actual instances can be created through a {@link #builder(QuerydslPredicateExecutor) builder} to query for
+ * {@link Builder#single()} or {@link Builder#many()} objects.
+ * Example:
+ * <pre class="code">
+ * interface BookRepository extends Repository&lt;Book, String&gt;, QuerydslPredicateExecutor&lt;Book&gt;{}
+ *
+ * BookRepository repository = …;
+ * TypeRuntimeWiring wiring = …;
+ *
+ * wiring.dataFetcher("books", QuerydslDataFetcherSupport.builder(repository).many())
+ *       .dataFetcher("book", QuerydslDataFetcherSupport.builder(repositories).single());
+ * </pre>
+ *
+ * @param <T> returned result type
+ * @author Mark Paluch
+ * @since 1.0.0
+ * @see QuerydslPredicateExecutor
+ * @see Predicate
+ * @see QuerydslBinderCustomizer
+ */
+public abstract class QuerydslDataFetcherSupport<T> {
```

Comment:
Technically it is a "support" class that doesn't implement `DataFetcher`, but it is also the main entry point to create Querydsl `DataFetcher`'s and having that on a support class is less than ideal. Actually the fact that it is a support class is somewhat of an implementation detail that is not relevant to users of the class whose only public API is the static `builder` method.

One idea is to call it `QuerydslDataFetcher` and declare it as such:
```java
public abstract class QuerydslDataFetcher<T, R> implements DataFetcher<R> {
   // ...
}
```
It's still a support class though with many of the same issues but it reads better:
```
QuerydslDataFetcher.builder(...).single();
```
Or it could be turned into `QuerydslDataFetcherBuilder` with a `from(QuerydslPredicateExecutor)` static, factory method that returns an instance of the `QuerydslDataFetcherBuilder` and everything else, including the "support" class becomes private and hidden like the actual implementations. I think I prefer this approach unless there are other concerns or potential future possible additions that I'm not seeing.

Reply:
Let's go for now with `QuerydslDataFetcher.builder(...).single()`.